### ISSUE TITLE
ever-increasing memory footprint running dns-fuzz script

### ISSUE THREAD

**Issue by dhduvall** [nmap/nmap#1245]
I was running the `dns-fuzz` script to test a DNS server, with the arguments as suggested in the script:

    nmap -sU --script dns-fuzz --script-args timelimit=2h <IP-address>

On my limited-memory cloud instance, nmap's RSS grew to ~800MB before starting to run parts of the system out of memory. I tried again with more memory, and it lasted longer, but still grew more than the system could handle.

I've no idea whether this is a memory leak in nmap, or the Lua interface, or the script itself, but I would expect it to stabilize at some much smaller size.

This was on Debian stretch, version 7.40, but I saw the same behavior on MacOS 10.13, with nmap 7.70 from homebrew.

**Comment by dmiller-nmap** [comment]
I took a glance at `dns-fuzz` and I can't see anything in the script itself that would be growing. It runs a loop that builds a string, modifies the string, sends a packet, gets a response, and repeats. No results are collected in a table or anything like I expected. Sounds more like a memory leak somewhere.

Can you provide the output of `nmap --version`  for the affected systems? Thanks!

**Comment by dhduvall** [comment]
```
Nmap version 7.40 ( https://nmap.org )
Platform: x86_64-pc-linux-gnu
Compiled with: liblua-5.3.3 openssl-1.1.0c libpcre-8.39 libpcap-1.8.1 nmap-libdnet-1.12 ipv6
Compiled without:
Available nsock engines: epoll poll select
```

```
Nmap version 7.70 ( https://nmap.org )
Platform: x86_64-apple-darwin17.3.0
Compiled with: liblua-5.3.3 openssl-1.0.2n nmap-libssh2-1.8.0 libz-1.2.11 nmap-libpcre-7.6 nmap-libpcap-1.7.3 nmap-libdnet-1.12 ipv6
Compiled without:
Available nsock engines: kqueue poll select
```

Thanks!

**Comment by dmiller-nmap** [comment]
It's not supposed to "use up" sockets like that; they should be being garbage-collected. I did find that the `comm.setup_connect` function (used internally for all comm.lua functions) failed to `close()` the socket if the `connect` call failed. I think this could lead to leaked Nsock IODs ("sockets") and other memory "leaks".

I pushed a change in r37340 to address this. If it solves this problem, then it's worth hunting for other cases where `sock:close()` isn't called on error, especially for long-running scripts.

**Comment by nnposter** [comment]
Yesterday I have already eliminated the missing `sock:close()` as a possible culprit because `sock:connect` is not failing even once throughout the script execution.

Even if it were, as part of troubleshooting, I was forcing GC at regular intervals but this had no material effect. From [nsedoc](https://nmap.org/nsedoc/lib/nmap.html#close):

> Sockets are subject to garbage collection. Should you forget to close a socket, it will get closed before it gets deleted (on the next occasion Lua's garbage collector is run). However since garbage collection cycles are difficult to predict, it is considered good practice to close opened sockets.

**Comment by gvanem** [comment]
There are tons of leaks all over Nmap. I compiled Nmap.exe using [Visual Leak Detector](https://github.com/KindDragon/vld) on MSVC-2019 (Win-10).

A simple `nmap.exe -d -p80 -sT --release-memory 10.0.0.1` shows approx 54.000 leaks!
Half of them are related to this:
```
---------- Block 341 at 0x00ADDA30: 72 bytes ----------
  Leak Hash: 0x61CADB90, Count: 1, Total 72 bytes
  Call Stack (TID 5484):
    ucrtbase.dll!malloc()
    d:\agent\_work\4\s\src\vctools\crt\vcstartup\src\heap\new_scalar.cpp (35): nmap.exe!operator new() + 0x8 bytes
    F:\MingW32\src\inet\nmap\services.cc (313): nmap.exe!nmap_services_init() + 0x84 bytes
    F:\MingW32\src\inet\nmap\services.cc (449): nmap.exe!gettoppts() + 0xD bytes
  Data:
    A0 3C 29 0B    88 43 2B 0C    A0 3C 29 0B    01 00 AD 00     .<)..C+. .<).....
    03 00 00 00    74 63 70 00    C4 E0 77 00    A8 9C DA 01     ....tcp. ..w.....
    BC E4 77 00    03 00 00 00    0F 00 00 00    A8 69 AD 00     ..w..... .....i..
    98 5C 2A 0C    00 00 00 00    03 00 FF FF    A8 5C 2A 0C     .\*..... .....\*.
    A0 A8 6C 58    53 59 54 3F                                   ..lXSYT? ........
```

**Comment by fyodor** [comment]
Thanks for the report and discussion.  We're closing this for now as stale, but happy to re-open and investigate if anyone can reproduce with current Nmap.  In that case please either create a new issue and reference this one.
